consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+
+	for i, val := range pcs.Validators {
+		valCopy[i] = &PrioritizedValidator{Metadata: val.Metadata.Copy(), ProposerPriority: val.ProposerPriority}
+
+		if pcs.Proposer != nil && pcs.Proposer.Metadata.Address == val.Metadata.Address {
+			proposer = valCopy[i]
+		}
+	}
+
+	return &ProposerSnapshot{
+		Validators: valCopy,
+		Height:     pcs.Height,
+		Round:      pcs.Round,
+		Proposer:   proposer,
+	}
+}
+
+// ProposerCalculator interface - proposer calculator algorithm should implement this interface
+type ProposerCalculator interface {
+	// CalcProposer calculates next proposer
+	CalcProposer(round, height uint64) (types.Address, error)
+
+	// Update updates ProposerSnapshot to block with number `blockNumber`
+	Update(blockNumber uint64, config *runtimeConfig, state *State) error
+
+	// GetLatestProposer returns latest calculated proposer if any
+	GetLatestProposer(round, height uint64) (types.Address, bool)
+
+	// GetSnapshot returns copy of the current snapshot
+	GetSnapshot() (*ProposerSnapshot, bool)
+}
+
+// check if interface is matching implementation
+var _ ProposerCalculator = (*proposerCalculator)(nil)
+
+type proposerCalculator struct {
+	// current snapshot
+	snapshot *ProposerSnapshot
+
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculator(config *runtimeConfig, logger hclog.Logger) (*proposerCalculator, error) {
+	snap, err := NewProposerSnapshotFromState(config, logger)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &proposerCalculator{
+		snapshot: snap,
+		logger:   logger,
+	}, nil
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculatorFromSnapshot(pcs *ProposerSnapshot, logger hclog.Logger) *proposerCalculator {
+	return &proposerCalculator{
+		snapshot: pcs.Copy(),
+		logger:   logger,
+	}
+}
+
+// Get copy of the proposers' snapshot
+func (pc *proposerCalculator) GetSnapshot() (*ProposerSnapshot, bool) {
+	if pc.snapshot == nil {
+		return nil, false
+	}
+
+	return pc.snapshot.Copy(), true
+}
+
+// GetLatestProposer returns latest calculated proposer if any
+func (pc *proposerCalculator) GetLatestProposer(round, height uint64) (types.Address, bool) {
+	// round must be same as saved one and proposer must exist
+	if pc.snapshot == nil || pc.snapshot.Proposer == nil || pc.snapshot.Round != round || pc.snapshot.Height != height {
+		pc.logger.Info("Get latest proposer not found", "height", height, "round", round,
+			"pc height", pc.snapshot.Height, "pc round", pc.snapshot.Round)
+
+		return types.ZeroAddress, false
+	}
+
+	pc.logger.Info("Get latest proposer",
+		"height", height, "round", round, "address", pc.snapshot.Proposer.Metadata.Address)
+
+	return pc.snapshot.Proposer.Metadata.Address, true
+}
+
+// CalcProposer calculates next proposer
+func (pc *proposerCalculator) CalcProposer(round, height uint64) (types.Address, error) {
```

Comment:
```suggestion
func (pc *proposerCalculator) CalcProposer(round uint64) (types.Address, error) {
```
The Snapshot here should already know what the height of the block is.

Reply:
Yep, it should. Since we receive both `round` and `height` from `go-ibft` when `CalcProposer` is called, we use the received `height` just to do an additional validation before actually calculating proposer. In this way we are sure that `ProposerSnapshot` data is completely aligned with the state of `go-ibft`. For now, it can detect possible bugs/race conditions in saving/loading `ProposerSnapshot` feature.